logic/concludable/ConjunctionConcludable.java

```diff
@@ -112,21 +118,107 @@ public Value asValue() {
 
         public Relation(final RelationConstraint constraint) {
             super(copyConstraint(constraint));
         }
 
         @Override
-        public Stream<Unification> unify(ThenConcludable.Relation unifyWith) {
-            // Check the relation variables' isa constraint labels and prune if there is no intersection
-
-            // Find all roleplayer mapping combinations, which should have the form:
-            // Set<List<Pair<RelationConstraint.RolePlayer, RelationConstraint.RolePlayer>>> rolePlayerMappings
-            // For each, prune if there is no label intersection (or any other pruning, e.g. by value)
-            // Then build a Unification for each valid combination
-            return Stream.empty(); // TODO
+        public Stream<Map<Reference, Set<Reference>>> unify(ThenConcludable.Relation unifyWith) {
+            if (this.constraint().players().size() > unifyWith.constraint().players().size()) return Stream.empty();
+            Map<Reference, Set<Reference>> variableUnifier = new HashMap<>();
+            Optional<Map<Reference, Set<Reference>>> newUnifier;
+            if (constraint.owner().reference().isName()) {
+                newUnifier = tryExtendUnifier(this.constraint().owner(), unifyWith.constraint().owner(), variableUnifier);
+                if (newUnifier.isPresent()) variableUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            if (constraint.owner().isa().isPresent() && constraint.owner().isa().get().type().reference().isName()) {
+                assert unifyWith.constraint().owner().isa().isPresent();
+                newUnifier = tryExtendUnifier(this.constraint().owner().isa().get().type(),
+                        unifyWith.constraint().owner().isa().get().type(), variableUnifier);
+                if (newUnifier.isPresent()) variableUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            Map<Reference, Set<Reference>> finalVariableUnifier = variableUnifier;
+            List<RolePlayer> conjRolePlayers = Collections.unmodifiableList(constraint.players());
+            List<RolePlayer> thenRolePlayers = Collections.unmodifiableList(unifyWith.constraint().players());
+            return unifyRP(conjRolePlayers, thenRolePlayers, new HashMap<>(), 0)
+                    .map(unifier -> convertRolePlayerUnifier(unifier, thenRolePlayers, finalVariableUnifier));
+        }
+
+        private Stream<Map<RolePlayer, Set<Integer>>> unifyRP(
+                List<RolePlayer> conjRolePlayers, List<RolePlayer> thenRolePlayers,
+                Map<RolePlayer, Set<Integer>> unifier, Integer startCon) {
+
+            if (startCon == conjRolePlayers.size()) return Stream.of(unifier);
+            RolePlayer conjRP = conjRolePlayers.get(startCon);
+
+            return IntStream.range(0, thenRolePlayers.size())
```

Comment:
i think i see what you're doing, and i think using index is ok. I think it can be made marginally cleaner by doing the following:
make a small intermediate method called `withRolePlayers(constraint.players(), unifyWith.constraint().players(), finalVariableUnifier)`
this method internally does:
```
withRolePlayers(players, otherPlayers, baseUnifier) {
Stream<> rolePlayerMappingIndices =  unifyRP(...)
return rolePlayerMappingIndices.map(indexMap -> unifierWithRolePlayers(indexMap, baseUnifier))
}
```
this way the naming helps the reader understand what is going on with the numbers

Reply:
Indexing was not my first choice, but all other solutions I came up with didn't work because it kept stumbling on the fact that the `equals` method was too loose, while `==` was too strict (it would see RolePlayers that should have been the same as distinct). I agree that it makes the code more opaque, unfortunately.

Reply:
I don't I see the benefit of this intermediate function. I sympathise, however, with the story being unclear, so I've made a number of renamings. The `UnifyRP` is now called `matchRolePlayerIndices`. This is much more accurate to what it actually does. Now it's clear it doesn't return a unifier, just an indices matching. Now we take this index mapping and map it to a unifier. So now the code reads

```
            return matchRolePlayerIndices(conjRolePlayers, thenRolePlayers, new HashMap<>())
                    .map(indexMap -> mapRolePlayerMatchingToUnifier(indexMap, thenRolePlayers, baseVariableUnifier));
```

Reply:
The main problem now is that we are slightly two-faced, because the intermediate Stream is from RolePlayers to Set<Indices>. It would feel more consistent to do a mapping from Indiceis -> Set<Indices>. However, this will add more complexity to the code. I therefore think that the current solution a good tradeoff, but let me know what you think.

Reply:
There was one other solution I thought of, which is to wrap the RolePlayers into an Object so that they also contained their Index. I thought this was a little over elaborate, but if the current implementation is too terse, I will refactor it with that idea.